legacy/firmware/ethereum.c

```diff
@@ -976,23 +976,34 @@ int ethereum_message_verify(const EthereumVerifyMessage *msg) {
   return 0;
 }
 
+/*
+ * EIP-712 hashes might have no message_hash if primaryType="EIP712Domain".
+ * In this case, set domain_only_hash=true.
+ */
 static void ethereum_typed_hash(const uint8_t domain_separator_hash[32],
                                 const uint8_t message_hash[32],
-                                uint8_t hash[32]) {
+                                bool domain_only_hash, uint8_t hash[32]) {
   struct SHA3_CTX ctx = {0};
   sha3_256_Init(&ctx);
   sha3_Update(&ctx, (const uint8_t *)"\x19\x01", 2);
   sha3_Update(&ctx, domain_separator_hash, 32);
-  sha3_Update(&ctx, message_hash, 32);
+  if (!domain_only_hash) {
+    sha3_Update(&ctx, message_hash, 32);
+  }
   keccak_Final(&ctx, hash);
 }
 
 void ethereum_typed_hash_sign(const EthereumSignTypedHash *msg,
                               const HDNode *node,
                               EthereumTypedDataSignature *resp) {
   uint8_t hash[32] = {0};
-  ethereum_typed_hash(msg->domain_separator_hash.bytes, msg->message_hash.bytes,
-                      hash);
+
+  if (msg->has_message_hash) {
```

Comment:
i believe you don't need this `if`. The field `message_hash.bytes` still exists and is of the right type

so just pass `msg->has_message_hash` as the 3rd argument

Reply:
Done in https://github.com/trezor/trezor-firmware/pull/2036/commits/34b0008275b4fcfbad801e4ffa4416fbe561a772 and I managed to confirm it works locally :smile: